collector/lib/ConfigLoader.h

```diff
@@ -1,14 +1,167 @@
 #ifndef _CONFIG_LOADER_H_
 #define _CONFIG_LOADER_H_
 
+#include <optional>
+
+#include <gtest/gtest_prod.h>
 #include <yaml-cpp/yaml.h>
 
+#include "internalapi/sensor/collector.pb.h"
+
 #include "CollectorConfig.h"
 #include "Inotify.h"
 #include "StoppableThread.h"
 
 namespace collector {
 
+class ParserError {
+ public:
+  ParserError() = default;
+  ParserError(ParserError&&) noexcept = default;
+  ParserError(const ParserError&) = default;
+  ParserError& operator=(const ParserError&) = default;
+  ParserError& operator=(ParserError&&) noexcept = default;
+  ~ParserError() = default;
+
+  ParserError(const char* msg) : msg_(msg) {}
+  ParserError(std::string msg) : msg_(std::move(msg)) {}
+
+  const std::string& What() const { return msg_; }
+
+  template <typename T>
+  friend ParserError& operator<<(ParserError& e, const T msg) {
+    std::stringstream ss;
+    ss << msg;
+    e.msg_ += ss.str();
+    return e;
+  }
+
+  friend std::ostream& operator<<(std::ostream& os, const ParserError& err) {
+    os << err.What();
+    return os;
+  }
+
+  friend bool operator==(const ParserError& lhs, const ParserError& rhs) { return lhs.msg_ == rhs.msg_; }
+
+ private:
+  std::string msg_;
+};
+
+using ParserResult = std::optional<std::vector<ParserError>>;
```

Comment:
After going through the code I still don't like this. I think it is simpler to not have optional. You have some if statements checking the optional, before inserting one list of errors into another. This check is not necessary if you just use a vector. Also I think there is only one place where `empty()` is needed if optional is not used. I created a branch where I made this change and the other changes that need to be made to get everything to work with just the plain vector. Please take a look and merge it into your branch if you like it. https://github.com/stackrox/collector/pull/2042

Reply:
Answered in the PR.

TL;DR the way used in this PR is more idiomatic C++17